src/box/lua/experimental/connpool.lua

```diff
@@ -54,6 +54,45 @@ local function connect_using_config(instance_name, opts)
     return conn
 end
 
+local function is_labels_match(labels, instance_labels)
+    if labels == nil then
+        return true
+    end
+    if instance_labels == nil then
+        return false
+    end
+    for _, label in pairs(labels) do
+        if instance_labels[label] == nil then
+            return false
+        end
+    end
+    return true
+end
+
+local function get_candidates(role_name, opts)
+    checks('string', {
+        labels = '?table',
+    })
+    opts = opts or {}
+
+    local candidates = {}
+    for instance_name in pairs(config:instances()) do
+        local iconfig = config:get('', {instance = instance_name})
+        assert(iconfig ~= nil and type(iconfig) == 'table')
+        local roles = iconfig.roles or {}
+        if is_labels_match(opts.labels, iconfig.labels) then
+            for _, role in pairs(roles) do
+                if role == role_name then
+                    table.insert(candidates, instance_name)
+                    break
+                end
+            end
+        end
+    end
```

Comment:
I would structurize the code in a way that allows to decouple the filters. For example,

```lua
local function filter_by_role(instance_name)
    if opts.role == nil then
        return true
    end
    local roles = config:get('roles', {instance = instance_name})
    return table_has_key(roles, opts.role)
end

local function filter_by_labels(instance_name)
    if opts.labels == nil then
        return true
    end
    local labels = config:get('labels', {instance = instance_name})
    return table_is_covered_by(labels, opts.labels)
end

return fun.iter(config:instances())
    :filter(filter_by_role)
    :filter(filter_by_labels)
    :totable()
```

(Extracting the `table_{has/is}_*` helpers would also increase readability, IMHO.)

(I don't test this code, it is just illustration.)

I don't think that repeated `config:get()` call is a problem. If it is, it is better to implement some caching inside the call.

Reply:
Reworked this part.